### DISCUSSION TITLE
Shepard's distortion

### DISCUSSION THREAD

**Discussion by tourtiere** [libvips/libvips#2352] [Show and tell]
Hi,
Just a quick note to say that I previously used ImageMagick and its shepard command to distort large orthophoto mosaics, but it was using too much RAM. I found libvips and its mapim command so I came up with this script and it really improved my project: https://github.com/tourtiere/light-distortion. Feel free to use it as you want in the examples.

Libvips is very powerful and fun to use. Thank you John and other contributors for this incredible work, have a nice day to you all

**Comment by jcupitt** [comment]
Hi @tourtiere,

That's very cool! I didn't realize Shepard was so straightforward. You can make your loop very slightly faster with:

```python
    for p1, p2 in couples:

        diff = index - p2

        distance = (diff[0]**2 + diff[1]**2)
        distance = distance.ifthenelse(distance, 0.1)

        weight = 1.0 / distance

        delta = [(p1[0] - p2[0]), (p1[1] - p2[1])] * weight

        weights.append(weight)
        deltas.append(delta)

    # add, normalize
    index += pyvips.Image.sum(deltas) / pyvips.Image.sum(weights)
```

Though it doesn't make a lot of difference.

I added a (1/x) shortcut to pow in git master libvips which helps a little too.

**Reply by jcupitt** [reply]
1/x patch: https://github.com/libvips/libvips/commit/dd4b11063a19be076cf9c89000da7d9451dfc2b9

**Reply by jcupitt** [reply]
I added your nice code to the pyvips examples area, I hope that's OK.

https://github.com/libvips/pyvips/commit/1b0fb4c0a7a636d24ce39f1a19ca8738101c244e

**Reply by tourtiere** [reply]
It's perfect! Thank you
